adapters/repos/db/search.go

```diff
@@ -176,6 +179,71 @@ func hydrateObjectsIntoSearchResults(objs []*storobj.Object, dists []float32) se
 	return res
 }
 
+func (d *DB) ReferenceVectorSearch(ctx context.Context,
+	obj *models.Object, refProps map[string]struct{},
+) ([][]float32, error) {
+	return d.getReferenceVectorsFromParent(ctx, obj, refProps)
+}
+
+func (d *DB) getReferenceVectorsFromParent(ctx context.Context,
+	parent *models.Object, referencePropNames map[string]struct{},
+) (refVecs [][]float32, err error) {
+	props := parent.Properties.(map[string]interface{})
+
+	// use the ids from parent's beacons to find the referenced objects
+	beacons := beaconsForVectorization(props, referencePropNames)
+	for _, beacon := range beacons {
+		res, findErr := d.findReferenceObject(ctx, beacon)
+		if findErr != nil {
+			err = findErr
+			return
+		}
+
+		// if the ref'd object has a vector, we grab it.
+		// these will be used to compute the parent's
+		// vector eventually
+		if res.Vector != nil {
+			refVecs = append(refVecs, res.Vector)
+		}
+	}
+
+	return
+}
+
+func beaconsForVectorization(allProps map[string]interface{},
+	targetRefProps map[string]struct{},
+) []string {
+	var beacons []string
+
+	// add any refs that were supplied as a part of the parent
+	// object, like when caller is AddObject/UpdateObject
+	for prop, val := range allProps {
+		if _, ok := targetRefProps[prop]; ok {
+			refs := val.(models.MultipleRef)
+			for _, ref := range refs {
+				beacons = append(beacons, ref.Beacon.String())
+			}
+		}
+	}
+
+	return beacons
+}
+
+func (d *DB) findReferenceObject(ctx context.Context, beacon string) (res *search.Result, err error) {
+	rest, id := path.Split(beacon)
+	_, class := path.Split(path.Clean(rest))
+
+	res, err = d.Object(ctx, class, strfmt.UUID(id),
+		search.SelectProperties{}, additional.Properties{})
+	if err != nil || res == nil {
+		if err == nil {
+			err = fmt.Errorf("not found")
+		}
+		err = fmt.Errorf("find object with beacon %q': %w", beacon, err)
+	}
```

Comment:
Please take into account backward compatibility of beacon format, they all now should have class inside but sometimes `ref.Class == ""` is empty, in that case please use `d.ObjectByID` method to find the object